AX: Merge AccessibilitySVGRoot into AccessibilitySVGObject

hoffmanjoshua opened the pull request:
#### f9fb9507fd1309f85837044a121c7777fd025958
<pre>
AX: Merge AccessibilitySVGRoot into AccessibilitySVGObject
<a href="https://bugs.webkit.org/show_bug.cgi?id=296699">https://bugs.webkit.org/show_bug.cgi?id=296699</a>
<a href="https://rdar.apple.com/157114857">rdar://157114857</a>

Reviewed by Tyler Wilcock.

We can represent SVG roots as SVG objects after bringing over the logic from
AccessibilitySVGRoot.

This PR also includes a few Unified Build fixes that were uncovered after removing
AccessibilitySVGRoot from the sources list.

* Source/WebCore/Sources.txt:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::createObjectFromRenderer):
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
* Source/WebCore/accessibility/AccessibilityObject.h:
(WebCore::AccessibilityObject::isAccessibilitySVGObjectInstance const):
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::determineAccessibilityRole):
(WebCore::AccessibilityRenderObject::remoteSVGRootElement const):
(WebCore::AccessibilityRenderObject::addRemoteSVGChildren):
* Source/WebCore/accessibility/AccessibilityRenderObject.h:
* Source/WebCore/accessibility/AccessibilitySVGObject.cpp:
(WebCore::AccessibilitySVGObject::AccessibilitySVGObject):
(WebCore::AccessibilitySVGObject::create):
(WebCore::AccessibilitySVGObject::determineAccessibilityRole):
(WebCore::AccessibilitySVGObject::parentObject const):
(WebCore::AccessibilitySVGObject::isRootWithAccessibleContent const):
* Source/WebCore/accessibility/AccessibilitySVGObject.h:
(WebCore::AccessibilitySVGObject::setParent):
(isType):
* Source/WebCore/accessibility/AccessibilitySVGRoot.cpp: Removed.
* Source/WebCore/accessibility/AccessibilitySVGRoot.h: Removed.
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::isDefaultValue):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:

Canonical link: <a href="https://commits.webkit.org/298044@main">https://commits.webkit.org/298044@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7f335c7da67eedf9ac1d52d191cc5c0613732b61

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/114030 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/33782 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/24240 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/120200 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/64805 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/115919 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/34410 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/42343 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/86669 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/41677 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/116978 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/27392 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/102413 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/67052 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/26577 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/20544 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/63905 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/96757 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/20661 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/123427 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/41063 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/30586 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/95500 "Found 2 new test failures: imported/w3c/web-platform-tests/html/semantics/forms/constraints/infinite_backtracking.html inspector/dom-debugger/event-animation-frame-breakpoints.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/41445 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/98625 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/95281 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24289 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/40416 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/18223 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/37186 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/40938 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/46440 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/40566 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/43865 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/42320 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/AccessibilitySVGObject.cpp:
```diff
@@ -319,4 +327,59 @@ AccessibilityRole AccessibilitySVGObject::determineAccessibilityRole()
+    return AccessibilityRenderObject::parentObject();
+}
+
+bool AccessibilitySVGObject::hasAccessibleContent() const
```
Maybe this method name should change now that it's on the more general `AccessibilitySVGObject` class. Maybe `isRootWithAccessibleContent()`?

hoffmanjoshua replied:
Updated!

twilco commented on Source/WebCore/accessibility/AccessibilitySVGObject.cpp:
```diff
@@ -319,4 +327,59 @@ AccessibilityRole AccessibilitySVGObject::determineAccessibilityRole()
+    if (m_parent)
+        return m_parent.get();
+
+    return AccessibilityRenderObject::parentObject();
```
While we're here, maybe we can clean up this comment? I personally think it's easier to read comments close to the actual code they're talking about, and for comments about branches to be inside the scope of the branch, both of which don't happen here. So maybe something like:

```cpp
if (m_parent) {
    // If a parent was set because this is a remote SVG resource, use that.
    return m_parent.get();
}
// Otherwise, we should rely on the standard render tree for the parent.
return AccessibilityRenderObject::parentObject();
```

Also should we be checking `if (m_isRoot && m_parent)`? We only want this behavior for the root, right? Probably we need to keep this kind of thing in mind as we shrink classes into each other.

hoffmanjoshua replied:
>Also should we be checking if (m_isRoot && m_parent)? We only want this behavior for the root, right? Probably we need to keep this kind of thing in mind as we shrink classes into each other.

I decided to add an ASSERT(m_isSVGRoot) instead, inside the m_parent block. If we've gotten in there, something has gone wrong.

twilco commented on Source/WebCore/accessibility/AccessibilitySVGObject.cpp:
```diff
@@ -319,4 +327,59 @@ AccessibilityRole AccessibilitySVGObject::determineAccessibilityRole()
+void AccessibilitySVGObject::setParent(AccessibilityRenderObject* parent)
+{
+    m_parent = parent;
+}
```
Should we `ASSERT(isRoot)`? This shouldn't happen for non-root things, right? Also, maybe worth defining this short method in the header to help the compiler inline the call?

hoffmanjoshua replied:
Done!

twilco commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
```diff
@@ -97,7 +97,7 @@ String AXIsolatedObject::debugDescriptionInternal(bool verbose, std::optional<Op
 }
 
-static bool isDefaultValue(AXProperty property, AXPropertyValueVariant& value)
+bool isDefaultValue(AXProperty property, AXPropertyValueVariant& value)
```
I'm confused how this ever compiled with `static`...

hoffmanjoshua replied:
Some unified build magic it seems 🤷🏻‍♂️

hoffmanjoshua committed:
AX: Merge AccessibilitySVGRoot into AccessibilitySVGObject
https://bugs.webkit.org/show_bug.cgi?id=296699
rdar://157114857

Reviewed by Tyler Wilcock.

We can represent SVG roots as SVG objects after bringing over the logic from
AccessibilitySVGRoot.

This PR also includes a few Unified Build fixes that were uncovered after removing
AccessibilitySVGRoot from the sources list.

* Source/WebCore/Sources.txt:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::createObjectFromRenderer):
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
* Source/WebCore/accessibility/AccessibilityObject.h:
(WebCore::AccessibilityObject::isAccessibilitySVGObjectInstance const):
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::determineAccessibilityRole):
(WebCore::AccessibilityRenderObject::remoteSVGRootElement const):
(WebCore::AccessibilityRenderObject::addRemoteSVGChildren):
* Source/WebCore/accessibility/AccessibilityRenderObject.h:
* Source/WebCore/accessibility/AccessibilitySVGObject.cpp:
(WebCore::AccessibilitySVGObject::AccessibilitySVGObject):
(WebCore::AccessibilitySVGObject::create):
(WebCore::AccessibilitySVGObject::determineAccessibilityRole):
(WebCore::AccessibilitySVGObject::parentObject const):
(WebCore::AccessibilitySVGObject::isRootWithAccessibleContent const):
* Source/WebCore/accessibility/AccessibilitySVGObject.h:
(WebCore::AccessibilitySVGObject::setParent):
(isType):
* Source/WebCore/accessibility/AccessibilitySVGRoot.cpp: Removed.
* Source/WebCore/accessibility/AccessibilitySVGRoot.h: Removed.
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::isDefaultValue):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:

Canonical link: https://commits.webkit.org/298044@main
